presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java

```diff
@@ -173,7 +175,45 @@ public void addPage(Page page, int partitionId, Block partitionIds)
 
     public DataSize getEstimatedSize()
     {
-        return new DataSize(estimatedSize, Unit.BYTE);
+        return new DataSize(estimatedSize, BYTE);
+    }
+
+    public DataSize getExpectedCompactedSize()
+    {
+        long size = 0;
+        long elementsSize = (channels.length > 0) ? sizeOf(channels[0].elements()) : 0;
+        size += elementsSize * channels.length;
+        size += sizeOf(valueAddresses.elements());
+        for (ObjectArrayList<Block> blocks : channels) {
+            for (Block block : blocks) {
+                size += block.getRetainedSizeInBytes();
+            }
+        }
+        return new DataSize(size, BYTE);
+    }
+
+    public void compact()
+    {
+        for (int channel = 0; channel < types.size(); channel++) {
+            Type type = types.get(channel);
+            ObjectArrayList<Block> blocks = channels[channel];
+            for (int i = nextBlockToCompact; i < blocks.size(); i++) {
+                Block block = blocks.get(i);
+                if (block.getSizeInBytes() < block.getRetainedSizeInBytes()) {
```

Comment:
Oops, I meant to remove this getExpectedCompactedSize()  before putting it up for review. Good catch! getRetainedSizeInBytes is the amount of heap memory retained by the Block, whereas getSizeInBytes is the logical size of the Block. This is particularly important when you build a Block that has a Slice inside it which has a length()  much smaller than the array it references